src/optimization/hooke_jeeves_method.cpp

```diff
@@ -0,0 +1,157 @@
+	communicator->send(target, k52::parallel::mpi::constants::kCommonTag, max_iteration_number_);
+	communicator->send(target, k52::parallel::mpi::constants::kCommonTag, init_step_);
+}
+
+void HookeJeevesMethod::Receive(boost::mpi::communicator* communicator, int source)
+{
+	communicator->recv(source, k52::parallel::mpi::constants::kCommonTag, acceleration_);
+	communicator->recv(source, k52::parallel::mpi::constants::kCommonTag, precision_);
+	communicator->recv(source, k52::parallel::mpi::constants::kCommonTag, max_iteration_number_);
+	communicator->recv(source, k52::parallel::mpi::constants::kCommonTag, init_step_);
+}
+#endif
+
+vector<double> HookeJeevesMethod::FindOptimalParameters(const vector<double>& initial_parameters)
+{
+	vector<double> arguments(initial_parameters);
+	vector<double> arguments_after_search(initial_parameters.size());
+	vector<double> steps_array = vector<double>(arguments.size(), init_step_);
+	size_t iteration = 0;
+
+	do
+	{
+		arguments_after_search = CoordinatewiseSearch(arguments, steps_array);
+		if (arguments == arguments_after_search)
+		{
+			for (size_t i = 0; i < steps_array.size(); ++i)
+			{
+				steps_array[i] /= 10;
+			}
+			continue;
+		}
+
+		vector<double> next_step_arguments(arguments.size());
+		vector<double> next_step_arguments_after_search(arguments.size());
+		while (true)
+		{
+			for (size_t i = 0; i<arguments.size(); ++i)
+			{
+				next_step_arguments[i] = arguments[i] + acceleration_
+					* (arguments_after_search[i] - arguments[i]);
+			}
+			next_step_arguments_after_search = CoordinatewiseSearch(next_step_arguments, steps_array);
+
+			arguments = arguments_after_search;
+			if (CountObjectiveFunctionValueToMinimize(arguments_after_search) <=
+					CountObjectiveFunctionValueToMinimize(next_step_arguments_after_search))
+			{
+				break;
+			}
+			arguments_after_search = next_step_arguments_after_search;
+
+			iteration++;
+			// Check number of iterations
+			if (iteration == max_iteration_number_)
+			{
+				std::cout << "Solution not found" << std::endl;
+				return arguments_after_search;
+			}
+		}		
+	} while (!IsExitCriteriaFulfilled(steps_array));
+	return arguments_after_search;
+}
+
+bool HookeJeevesMethod::IsExitCriteriaFulfilled(
+	const vector<double>& steps_array) const
+{
+	double temp = 0;
+	// check if step is small enough
+	for (size_t j = 0; j<steps_array.size(); ++j)
+	{
+		temp += steps_array[j] * steps_array[j];
+	}
+	if (std::sqrt(temp) <= precision_)
+	{
+		return true;
+	}
+	return false;
+}
+
+vector<double> HookeJeevesMethod::CoordinatewiseSearch(
+	const vector<double>& arguments,
+	const vector<double>& steps_array) const
+{
+	vector<double> new_arguments(arguments);
+	vector<double> temp(arguments);
+	for (size_t i = 0; i < arguments.size(); ++i)
+	{
+		temp[i] = new_arguments[i] + steps_array[i];
+		if (CountObjectiveFunctionValueToMinimize(temp) >= CountObjectiveFunctionValueToMinimize(new_arguments))
```

Comment:
CountObjectiveFunctionValueToMinimize may be reeeealy time consuming (hours to calculate).
So must not recalculate it with same arguments (e.g.  new_arguments not changed) but save the result in variable instead